### ISSUE TITLE
How to whitelist git tracked files

### ISSUE THREAD

**Issue by unphased** [sharkdp/fd#503]
I'm mainly dealing with the issue of dotfiles not showing up when `fd` is used with `fzf` (`export FZF_DEFAULT_COMMAND="fd --type file"`) to efficiently list files on disk. 

Clearly the default of not including hidden files is reasonable, but the one exception to this is dotfiles e.g. `.vimrc`, `~/.config/terminator/config`, `.profile` all files that I will actually edit quite often. 

There seem to be 2 approaches

- add to the `fd` invocation to explicitly include files from `git ls-files` somehow. It's unclear how to do this cleanly. It's unclear if it is a reasonable thing to ask `fd` itself to whitelist git tracked files for us.
- run `fd -H` and add many patterns `.svn` `.git` etc to the `.ignore`. This is probably sane but I don't know, it doesn't give me warm and fuzzies. 

Ideas?

**Comment by sharkdp** [comment]
You can create a `.fdignore` file with a whitelist:
```gitignore
!/.vimrc
!/.profile
!/.config
```

**Comment by tmccombs** [comment]
That works, but does require keeping the whitelist up to date with files check in to git

**Comment by unphased** [comment]
Yes the explicit whitelist is option 3 and the least maintainable of them all.

**Comment by numToStr** [comment]
I also have a similar scenario but the whitelist in `.fdignore` is not working for me.

```
!.config/nvim/init.vim
.config
```

When I search, the list does not show the `init.vim` file.

**Comment by unphased** [comment]
@numToStr Have you tried reversing the order?

**Comment by numToStr** [comment]
@unphased Yeah, I tried that. But no luck.

**Comment by unphased** [comment]
@numToStr Your problem may be a manifestation of this bug: https://github.com/BurntSushi/ripgrep/issues/1374

it may be possible that both projects are using the same rust lib for ignore file handling, and it has quite some issues. 

I'm unable to use my ignore file to whitelist my `.config/nvim/init.vim` file, basically, the same issue as yours. It simply hasnt bothered me till now because I don't edit that file much (it sources my main .vimrc file)

**Comment by numToStr** [comment]
@unphased Any idea when this gonna be fixed?

**Comment by unphased** [comment]
I came back here and realized after a while that I wrote this issue. 3 years later and no change it seems. Does someone know what the upstream dependency is which does the gitignore logic for `fd` and `rg`? It's a very obvious use case that a git-tracked hidden file checked into git should override ignoring hidden files. 

There should be two improvements
- git tracked files to be nonignored, with a high priority (preferably configurable)
- multiple levels of gitignore file patterns. it seems from my reading of issues that specifying unignored items inside `.fdignore` is supposed to take precedence over ignore patterns in `.ignore`, which could be considered sufficient. But again making this configurable would be highly preferred.

**Comment by tmccombs** [comment]
Does someone know what the upstream dependency is which does the gitignore logic for fd and rg?
 
 It is the [`ignore` crate](https://crates.io/crates/ignore). which is part of the ripgrep (rg) project.

**Comment by unphased** [comment]
Thanks. Alright, it seems not in scope to add git ls-files non-ignore to that if it's just for parsing gitignore syntax essentially.

**Comment by unphased** [comment]
But still i'd like to lightly suggest (since this is more in `fd-find`'s scope than it is `ripgrep`'s)...
- check if git executable is present in environment and search dir is git repo or in a git repo
- if so, use it to run git ls-files
- by default or if configured to do so: apply git tracked files as exceptions to ignore rules
